src/Symfony/Component/HttpClient/Response/HttplugPromise.php

```diff
@@ -0,0 +1,243 @@
+<?php
+
+/*
+ * This file is part of the Symfony package.
+ *
+ * (c) Fabien Potencier <fabien@symfony.com>
+ *
+ * For the full copyright and license information, please view the LICENSE
+ * file that was distributed with this source code.
+ */
+
+namespace Symfony\Component\HttpClient\Response;
+
+use GuzzleHttp\Promise\PromiseInterface;
+use Http\Client\Exception\NetworkException;
+use Http\Promise\Promise;
+use Psr\Http\Message\RequestInterface;
+use Psr\Http\Message\ResponseFactoryInterface;
+use Psr\Http\Message\ResponseInterface as Psr7ResponseInterface;
+use Psr\Http\Message\StreamFactoryInterface;
+use Symfony\Contracts\HttpClient\Exception\TransportExceptionInterface;
+use Symfony\Contracts\HttpClient\HttpClientInterface;
+use Symfony\Contracts\HttpClient\ResponseInterface;
+
+/**
+ * @author Tobias Nyholm <tobias.nyholm@gmail.com>
+ *
+ * @internal
+ */
+final class HttplugPromise implements Promise
+{
+    private $request;
+    private $response;
+    private $client;
+    private $responseFactory;
+    private $streamFactory;
+    private $result;
+    private $state = Promise::PENDING;
+    private $callbacks = [];
+
+    private static $promisePool;
+
+    /**
+     * @param ResponseInterface|\Exception $response
+     */
+    public function __construct(RequestInterface $request, $response, HttpClientInterface $client, ResponseFactoryInterface $responseFactory, StreamFactoryInterface $streamFactory)
+    {
+        if ($response instanceof \Exception) {
```

Comment:
to me, `Error` should not be handled by promises: they're developer mistakes and must bubble asap to the debugging tools.

Reply:
well, `guzzle/promises` and `react/promise` are both catching Throwable for rejections, so this actually hurts interoperability.

Reply:
If one wants interop at this level, one can wrap this promise into a e.g. a guzzle one, that would do the job. But swallowing thrown `Error` as is the default when no error callback is registered is terrible for DX so that cannot be the default.